### PR TITLE
refactor(api)!: json data for ContentNode

### PR DESCRIPTION
**Pull request by usu**
Playground for implementation of https://github.com/ecamp/ecamp3/issues/2633

This is a Proof-of-Concept and not meant for merging. Most things will not work, among others:
- Schema is migrated, but data is not --> modifying existing content will not work
- Tests don't run
- Print will not run
- Creating an activity will not work as it relies on proper category data (which is not migrated)

Things that should work:
- Creating a new category in an existing camp
- Adding/modifying/deleting the following content types in this new category:
  - [x] ColumnLayout
  - [x] Notes
  - [x] SafetyConcept
  - [x] Storyconcept
  - [x] Storyboard
  - [x] MultiSelect

### Open ToDos to make tis POC mergeable
- [ ] fix tests
- [ ] fix frontend
- [ ] fix printing
- [ ] migrate dev data (or start from scratch?)
- [ ] implement copyFromPrototype
- [ ] Avoid string comparisons in switch/case --> introduce enum for ContentTypeName or for dataSchema (replacement for contentType->entityClass). Verify if ContentType entity is still needed.

### Time measurement
(contentNodes timing not fully comparable, as data is not migrated)

Endpoint | Collection ms | Entity ms
-- | -- | --
activities | 224 | 84
contentNodes | 170 | 60
days | 69 | 58
categories | 81 | 64
contentTypes | 46 | 43
camps | 65 | 79
periods | 72 | 75
materialLists | 59 | 52
scheduleEntries | 87 | 68
materialItems | 64 | 56
activityResponsibles | 53 | 55
dayResponsibles | 54 | 55
campCollaborations | 92 | 61

### CODE REVIEW

File: api/src/DataPersister/ContentNodeDataPersister.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+    /**
+     * @param ContentNode $data
+     */
+    public function beforeCreate($data): ContentNode {
+        // set root from parent
+        $data->parent->addChild($data);
+        $data->parent->root->addRootDescendant($data);
+
+        switch ($data->getContentTypeName()) {
```
Comment (pmattmann) [inline api/src/DataPersister/ContentNodeDataPersister.php:32]: Idea:
Move Default-Data into ContentType->jsonConfig

File: api/src/DataPersister/ContentNodeDataPersister.php [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        }
+
+        return $data;
+    }
+
+    /**
+     * @param ContentNode $data
+     */
+    public function beforeUpdate($data): ContentNode {
+        switch ($data->getContentTypeName()) {
```
Comment (pmattmann) [inline api/src/DataPersister/ContentNodeDataPersister.php:72]: Idea:
Move HTMLFilter Field-List to ContentType->jsonConfig

File: api/src/Entity/ContentNode.php [hunk 3]
```diff
@@ -116,14 +150,52 @@ abstract class ContentNode extends BaseEntity implements BelongsToContentNodeTre
...
+        $this->rootDescendants = new ArrayCollection();
+    }
+
+    /**
+     * Return dynamic validation groups.
+     *
+     * @return string[]
+     */
+    public static function validationGroupsPost(self $contentNode) {
+        switch ($contentNode->getContentTypeName()) {
```
Comment (pmattmann) [inline api/src/Entity/ContentNode.php:169]: Idea:
move to ContentType->jsconConfig

File: api/src/Entity/ContentNode.php [hunk 4]
```diff
@@ -116,14 +150,52 @@ abstract class ContentNode extends BaseEntity implements BelongsToContentNodeTre
...
+        }
+    }
+
+    /**
+     * Return dynamic validation groups.
+     *
+     * @return string[]
+     */
+    public static function validationGroupsPatch(self $contentNode) {
+        switch ($contentNode->getContentTypeName()) {
```
Comment (pmattmann) [inline api/src/Entity/ContentNode.php:181]: Idea:
move to ContentType->jsconConfig

**Review by pmattmann — COMMENTED** [review]
Idea:
Move some of the code into ContenType->jsonConfig

**Comment by usu** [comment]
### Hybridmodell
- data property in ContentNode
- zusätzliche Daten aus der DB, welche auf die ContentNode zeigen (z.B. MaterialItems zeigen auf ContentNode, Kursziele analog umgesetzt)

### Hybridmodell 2
- ContentNode als abstract class belassen
- 1 Default-Ableitung mit json data property (kann für die meisten ContentTypen verwendet werden)
- erlaubt zusätzliche Ableitungen (z.B. für MaterialItems, MZB, etc.)

### Datenintegrität innerhalb data Property?
Beispiel: MZB Waypoints und "Legs" zwischen MZB Waypoints

**Pull request closed without merging**